invoice/src/main/java/org/killbill/billing/invoice/generator/FixedAndRecurringInvoiceItemGenerator.java

```diff
@@ -213,12 +213,15 @@ boolean isSameDayAndSameSubscription(final InvoiceItem prevComputedFixedItem, fi
     private List<InvoiceItem> processRecurringEvent(final UUID invoiceId, final UUID accountId, final BillingEvent thisEvent, @Nullable final BillingEvent nextEvent,
                                                     final LocalDate targetDate, final Currency currency,
                                                     final InvoiceItemGeneratorLogger invoiceItemGeneratorLogger,
-                                                    final Map<UUID, SubscriptionFutureNotificationDates> perSubscriptionFutureNotificationDate,
+                                                    final Map<UUID, SubscriptionFutureNotificationDates> perSubscriptionFutureNotificationDates,
                                                     final InternalCallContext internalCallContext) throws InvoiceApiException {
 
         try {
             final List<InvoiceItem> items = new ArrayList<InvoiceItem>();
-
+            final LocalDate thisEventEffectiveDate = internalCallContext.toLocalDate(thisEvent.getEffectiveDate());
+            if (thisEventEffectiveDate.compareTo(targetDate) > 0) {
```

Comment:
Is this an optimization (unrelated to the bug)?

Reply:
No, it's part of the big fix: You could have future events in the billing events (past `targetDate`). For instance if you have a future cancelation you want to exit quickly otherwise this will clear the `perSubscriptionFutureNotificationDates` map.